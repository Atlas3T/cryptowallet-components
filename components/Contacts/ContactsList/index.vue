<template>
  <section
    :class="{ isSearchingContacts: isSearchingContacts }"
    class="contacts-list"
  >
    <h1 class="section-h1">
      {{ $t('contacts') }}
    </h1>
    <div class="flex-scroll-area-wrapper">
      <q-scroll-area class="contacts-list-scroll-area">
        <div class="scroll-content">
          <ContactListItem
            v-for="contact in filteredContacts"
            :key="contact.displayName"
            :contact="contact"
            :click-item-action="clickItemAction"
          />
        </div>
      </q-scroll-area>
    </div>
  </section>
</template>

<script>
import { mapState } from 'vuex';
import ContactListItem from '@/components/Contacts/ContactListItem';

export default {
  name: 'ContactsList',
  components: {
    ContactListItem,
  },
  props: {
    clickItemAction: {
      type: String,
      required: false,
    },
  },
  computed: {
    ...mapState({
      isSearchingContacts: (state) => {
        return state.search.isSearchingContacts;
      },
      searchingContactsQueryString: (state) => {
        return state.search.searchingContactsQueryString;
      },
      contacts: (state) => {
        return state.contacts.contacts;
      },
    }),

    /**
     * Filters contact based on the search input value
     * @return {Array}
     */
    filteredContacts() {
      return this.contacts.filter((contact) => { return new RegExp(`^${this.searchingContactsQueryString}`, 'i').test(contact.displayName); });
    },
  },
};
</script>

<style scoped>
.section-h1 {
  color: white;
  font-size: 0.9rem;
  font-family: 'Montserrat-Light';
  background: #162f46;
  padding: 0.5rem 1rem;
  border-bottom: 1px solid #0a2338;
  letter-spacing: 0.05rem;
  text-transform: uppercase;
}

.flex-scroll-area-wrapper {
  flex: auto;
  flex-grow: 1;
  position: relative;
  overflow: hidden;
  height: 100%;
}

.contacts-list-scroll-area {
  flex-grow: 1;
  width: 100%;
  height: 100%;
  position: absolute;
}

.contacts-list {
  border-top: 2px solid #09233a;
  position: relative;
  background: #16324a;
  display: flex;
  flex-direction: column;
  transition: height 100ms ease-in-out;
  height: 100%;
}

.contacts-list.isSearchingContacts {
  height: 85vh;
}

.scroll-content {
  overflow: hidden;
  padding: 0 1rem;
}
</style>
