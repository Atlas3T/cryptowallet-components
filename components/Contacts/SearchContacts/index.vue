<template>
  <section class="search-contacts">
    <div class="flex-wrapper">
      <div class="search-input-wrapper">
        <q-input
          v-model="searchQueryString"
          inverted
          clearable
          float-label="Search contacts"
          color="blueish"
          @focus="isSearching(true)"
          @blur="isSearching(false)"
        />

        <div
          v-show="isSearchingContacts"
          class="cancel-search-wrapper"
        >
          <q-btn
            color="secondary"
            label="Cancel"
            @click="cancelSearch"
          />
        </div>
      </div>
    </div>
  </section>
</template>

<script>
import { mapState } from 'vuex';

export default {
  name: 'SearchContacts',
  data() {
    return {
    };
  },

  computed: {
    ...mapState({
      isSearchingContacts: (state) => {
        return state.search.isSearchingContacts;
      },
      searchingContactsQueryString: (state) => {
        return state.search.searchingContactsQueryString;
      },
    }),

    searchQueryString: {

      /**
       * Use a computed value getter to keep the input value in sync with the store
       * @return {String}
       */
      get() {
        return this.$store.state.search.searchingContactsQueryString;
      },

      /**
       * Use a computed value setter to update the store if input value changes
       * @param {String} value The new input value
       */
      set(value) {
        this.$store.dispatch('search/updateSearchContactsQueryString', { value });
      },
    },

  },

  methods: {

    /**
     * Updates the store search/isSearchingContacts state
     * @param  {Boolean} value Either is searching or isn't
     * @return {Boolean}
     */
    isSearching(value) {
      this.$store.dispatch('search/updateIsSearchingContacts', { value });
    },

    /**
     * Will be called if user clicks the CANCEL button while searching contacts
     * Will dispatch actions needed to reset the input value and reset the UI state
     */
    cancelSearch() {
      this.$store.dispatch('search/updateIsSearchingContacts', { value: false });
      this.$store.dispatch('search/updateSearchContactsQueryString', { value: '' });
    },
  },
};
</script>

<style>
.search-contacts {
  transition: height 250ms ease-in-out;
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 5em;
}

.search-input-wrapper {
  width: 90vw;
  display: flex;
}

.search-input-wrapper > .q-input {
  width: 60%;
  flex-grow: 1;
}

.cancel-search-wrapper > .q-btn {
  margin-left: 1rem;
  height: 100%;
}

.search-input-wrapper .column {
  padding: 0;
}
</style>
