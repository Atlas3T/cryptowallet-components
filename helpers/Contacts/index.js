import Vue from 'vue';

/**
 * Handles the contacts import and saves them in the store
 */
export default function Contacts() {
  this.vm = new Vue();
  this.import = (contacts) => {
    this.contacts = contacts;
    this.vm.$once('contactsImported', () => { return this.success(); });
    this.vm.$once('contactsImportFailure', () => { return this.failure(); });
    this.vm.$once('contactsImportError', (event) => { return this.error(event.payload); });

    try {
      const options = new ContactFindOptions();
      options.multiple = true;
      const fields = ['*'];
      navigator.contacts.find(fields, (contactsArray) => {
        this.contacts.contacts = contactsArray;
        this.vm.$emit('contactsImported');
      }, () => {
        this.vm.$emit('contactsImportFailure');
      }, options);
    } catch (err) {
      this.vm.$emit('contactsImportError', { payload: err });
    }

    return this;
  };

  this.on = (event, callback) => {
    this[event] = callback;
    return this;
  };
}
