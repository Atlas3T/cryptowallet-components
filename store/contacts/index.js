export default {
  namespaced: true,
  state: {
    contacts: [
      {
        id: 1,
        displayName: 'Terry Tibbs',
        address: 'ncousohfodicwjdclwdnc',
      },
    ],
  },
};
