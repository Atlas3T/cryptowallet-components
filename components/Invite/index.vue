<template>
  <div>
    <q-btn
      :label="$t('sendAppInvitation')"
      icon="send"
      color="primary"
      class="send-invitation"
      @click="sendAppInvitation"
    />

    <q-modal
      v-model="invitationModalOpened"
      class="dark-modal"
    >
      <div class="close-button-wrapper">
        <q-btn
          :label="$t('close')"
          color="secondary"
          size="sm"
          @click.prevent="invitationModalOpened = false"
        />
      </div>
      <SearchContacts />
      <ContactsList click-item-action="app-invitation" />
    </q-modal>

    <q-modal
      v-model="permissionsNoticeModalOpened"
      minimized
    >
      <div style="padding: 2rem">
        <div class="q-display-1 q-mb-md">
          {{ $t('permissions') }}
        </div>
        <p>{{ $t('contactsPermissions') }}</p>
        <q-btn
          color="primary"
          label="Ok"
          @click.prevent="closePermissionsNoticeModal"
        />
      </div>
    </q-modal>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import SearchContacts from '@/components/Contacts/SearchContacts';
import ContactsList from '@/components/Contacts/ContactsList';

export default {
  name: 'Invite',
  components: {
    SearchContacts,
    ContactsList,
  },

  data() {
    return {
      permissionsNoticeModalOpened: false,
      invitationModalOpened: false,
    };
  },
  computed: {
    ...mapState({
      delay: (state) => { return state.settings.delay; },
    }),
  },
  methods: {
    /**
     * Initiates the sending invitation process
     * If has premissions, just import contacts
     * If no premissions, show the premissions notice modal
     */
    sendAppInvitation() {
      const { permissions } = cordova.plugins;
      permissions.checkPermission(permissions.READ_CONTACTS, (status) => {
        if (status.hasPermission) {
          this.importContacts();
        } else {
          this.permissionsNoticeModalOpened = true;
        }
      }, () => {
        this.$toast.create(10, this.$t('contactsImportError'), this.delay.normal);
      });
    },

    /**
     * Closes the premissions notice modal and
     * initiaties contacts import
     */
    closePermissionsNoticeModal() {
      this.permissionsNoticeModalOpened = false;
      this.importContacts();
    },

    /**
     * Attempts to import and save contacts in the store
     */
    importContacts() {
      new this.ContactsImport(this)
        .on('success', () => {
          this.invitationModalOpened = true;
        })
        .on('failure', () => {
          this.$toast.create(420, this.$t('contactsImportFailure'), this.delay.normal);
        })
        .on('error', () => {
          this.$toast.create(10, this.$t('contactsImportError'), this.delay.normal);
        })
        .import(this.$store.state.contacts);
    },
  },
};
</script>

<style lang="scss">
  .close-button-wrapper {
    padding: 0.5rem;
    height: 2.7rem;
  }

  .dark-modal {
    .modal-content {
      background: #1e3c57;
    }
  }
</style>
