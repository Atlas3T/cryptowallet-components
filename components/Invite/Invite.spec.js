import { shallowMount, mount } from '@vue/test-utils';
import Invite from '@/components/Invite';
import SearchContacts from '@/components/Contacts/SearchContacts';
import ContactsList from '@/components/Contacts/ContactsList';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';
import cordovaMocks from '@/cordovaMocks';

describe('Invite', () => {
  let storeMocks;
  let wrapper;
  let store;

  function wrapperInit (options) {
    return shallowMount(Invite, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    wrapper = wrapperInit({ i18n, localVue, store: storeMocks.store });
    store = wrapper.vm.$store;
    cordovaMocks.initMocks();
  }

  beforeEach(() => storeInit());
  afterEach(() => cordovaMocks.destroyMocks());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders its child components', () => {
    expect(wrapper.contains(SearchContacts)).toBe(true);
    expect(wrapper.contains(ContactsList)).toBe(true);
  });

  it('renders two modals', () => {
    expect(wrapper.findAll('qmodal-stub').length === 2).toBe(true);
  });

  it('calls sendAppInvitation() method when you click on the button', () => {
    const sendAppInvitationMock = jest.fn();
    wrapper = mount(Invite, { i18n, localVue, store: storeMocks.store });
    wrapper.setMethods({ sendAppInvitation: sendAppInvitationMock });
    wrapper.find('.send-invitation').trigger('click');
    expect(sendAppInvitationMock).toHaveBeenCalled();
  });

  it('calls importContacts() method if you have READ_CONTACTS permissions', () => {
    const importContactsMock = jest.fn();
    wrapper.setMethods({ importContacts: importContactsMock });
    cordova.plugins.permissions.READ_CONTACTS.mockBehaviour = 'pass';
    wrapper.vm.sendAppInvitation();
    expect(importContactsMock).toHaveBeenCalled();
  });

  it('opens the permissionsNotice modal if you don\'t have READ_CONTACTS permissions', () => {
    expect(wrapper.vm.permissionsNoticeModalOpened).toBe(false);
    cordova.plugins.permissions.READ_CONTACTS.mockBehaviour = 'fail';
    wrapper.vm.sendAppInvitation();
    expect(wrapper.vm.permissionsNoticeModalOpened).toBe(true);
  });

  it('creates a toast on contact permissions error', () => {
    const createMock = jest.fn();
    wrapper.vm.$toast.create = createMock;
    cordova.plugins.permissions.READ_CONTACTS.mockBehaviour = 'error';
    wrapper.vm.sendAppInvitation();
    expect(createMock.mock.calls[0][1] === wrapper.vm.$t('contactsImportError')).toBe(true);
  });

  it('closes the permissionsNotice modal and calls importContacts() when closePermissionsNoticeModal() is called', () => {
    const importContactsMock = jest.fn();
    wrapper.setMethods({ importContacts: importContactsMock });
    wrapper.setData({ permissionsNoticeModalOpened: true });

    wrapper.vm.closePermissionsNoticeModal();
    expect(wrapper.vm.permissionsNoticeModalOpened).toBe(false);
    expect(importContactsMock).toHaveBeenCalled();
  });

  it('opens the invitationModal on importContacts() method success', () => {
    cordova.plugins.permissions.READ_CONTACTS.mockBehaviour = 'pass';
    navigator.contacts.mockBehaviour = 'pass';
    expect(wrapper.vm.invitationModalOpened).toBe(false);

    wrapper.vm.importContacts();
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.invitationModalOpened).toBe(true);
    });
  });

  it('creates a toast on importContacts() failure', () => {
    const createMock = jest.fn();
    wrapper.vm.$toast.create = createMock;
    cordova.plugins.permissions.READ_CONTACTS.mockBehaviour = 'pass';
    navigator.contacts.mockBehaviour = 'fail';
    wrapper.vm.importContacts();
    expect(createMock.mock.calls[0][1] === wrapper.vm.$t('contactsImportFailure')).toBe(true);
  });

  it('creates a toast on importContacts() error', () => {
    const createMock = jest.fn();
    wrapper.vm.$toast.create = createMock;
    cordova.plugins.permissions.READ_CONTACTS.mockBehaviour = 'pass';
    navigator.contacts.mockBehaviour = 'error';
    wrapper.vm.importContacts();
    expect(createMock.mock.calls[0][1] === wrapper.vm.$t('contactsImportError')).toBe(true);
  });
});
