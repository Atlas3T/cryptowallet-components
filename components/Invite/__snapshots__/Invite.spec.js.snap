// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invite renders and matches snapshot 1`] = `
<div>
  <qbtn-stub
    class="send-invitation"
    color="primary"
    icon="send"
    label="Invite to use the app"
  />
   
  <qmodal-stub
    class="dark-modal"
    position=""
    positionclasses="flex-center"
  >
    <div
      class="close-button-wrapper"
    >
      <qbtn-stub
        color="secondary"
        label="Close"
        size="sm"
      />
    </div>
     
    <searchcontacts-stub />
     
    <contactslist-stub
      clickitemaction="app-invitation"
    />
  </qmodal-stub>
   
  <qmodal-stub
    minimized="true"
    position=""
    positionclasses="flex-center"
  >
    <div
      style="padding: 2rem;"
    >
      <div
        class="q-display-1 q-mb-md"
      >
        Permissions
      </div>
       
      <p>
        We're going to ask you to give us access to your contacts. We have reasons buddy.
      </p>
       
      <qbtn-stub
        color="primary"
        label="Ok"
      />
    </div>
  </qmodal-stub>
</div>
`;
